exchanges/gateio/gateio.go

```diff
@@ -424,169 +1013,2938 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+	params := url.Values{}
+	if status < 1 || status > 3 {
+		return nil, fmt.Errorf("%s %v, only allowed status values are 1:failed, 2:borrowed, and 3:repayment", g.Name, errInvalidOrderStatus)
+	}
+	params.Set("status", strconv.Itoa(int(status)))
+	if !currency.IsEmpty() {
+		params.Set("currency", currency.String())
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	if offset > 0 {
+		params.Set("offset", strconv.FormatUint(offset, 10))
+	}
+	if reverse {
+		params.Set("reverse", strconv.FormatBool(reverse))
+	}
+	var response []CrossMarginLoanResponse
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginLoans, params, nil, &response)
+}
+
+// GetSingleBorrowLoanDetail retrieve single borrow loan detail
+func (g *Gateio) GetSingleBorrowLoanDetail(ctx context.Context, loanID string) (*CrossMarginLoanResponse, error) {
+	if loanID == "" {
+		return nil, errInvalidLoanID
+	}
+	var response CrossMarginLoanResponse
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginLoans+"/"+loanID, nil, nil, &response)
+}
+
+// *********************************Futures***************************************
+
+// GetAllFutureContracts  retrives list all futures contracts
+func (g *Gateio) GetAllFutureContracts(ctx context.Context, settle string) ([]FuturesContract, error) {
+	var contracts []FuturesContract
+	settle = strings.ToLower(settle)
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errMissingSettleCurrency
+	}
+	return contracts, g.SendHTTPRequest(ctx, exchange.RestSpot,
+		fmt.Sprintf(futuresSettleContracts, settle), &contracts)
+}
+
+// GetSingleContract returns a single contract info for the specified settle and Currency Pair (contract << in this case)
+func (g *Gateio) GetSingleContract(ctx context.Context, settle string, contract currency.Pair) (*FuturesContract, error) {
+	var futureContract FuturesContract
+	if contract.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	settle = strings.ToLower(settle)
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errMissingSettleCurrency
+	}
+	fPair, err := g.GetPairFormat(asset.Futures, true)
+	if err != nil {
+		fPair = currency.PairFormat{
+			Delimiter: currency.UnderscoreDelimiter,
+			Uppercase: true,
+		}
+	}
+	pairString := strings.ToUpper(fPair.Format(contract))
+	return &futureContract, g.SendHTTPRequest(ctx, exchange.RestSpot, fmt.Sprintf(futuresSettleContracts+"/%s", settle, pairString), &futureContract)
+}
+
+// GetFuturesOrderbook retrives futures order book data
+func (g *Gateio) GetFuturesOrderbook(ctx context.Context, settle string, contract currency.Pair, interval string, limit uint64, withOrderbookID bool) (*Orderbook, error) {
+	var response Orderbook
```

Comment:
Just another note here this doesn't need to get allocated until it gets used, in the event of an error. This can be adjust across this implementation.